utils/man/energymon-cmd-profile.1
.TH "ENERGYMON\-CMD\-PROFILE" "1" "2018-11-04" "energymon" "EnergyMon Utilities"
.SH "NAME"
.LP
energymon\-cmd\-profile \- profile the execution of a command
.SH "SYNPOSIS"
.LP
\fBenergymon\-cmd\-profile\fP
[\fIOPTION\fP]... \fICOMMAND\fP [\fIARG\fP...]
.SH "DESCRIPTION"
.LP
Prints time, energy, and average power for the execution of a command.
Uses the default EnergyMon implementation.
.SH "OPTIONS"
.LP
.TP
\fB\-h\fP, \fB\-\-help\fP
Prints the help screen.
.SH "EXAMPLES"
.TP
\fBenergymon\-cmd\-profile sleep 10\fP
Profile doing nothing for 10 seconds.
The \fBsleep\fP program is part of \fBGNU coreutils\fP.
.TP
\fBenergymon\-cmd\-profile stress \-c $(nproc) \-t 10\fP
Profile stressing the processor for 10 seconds.
The \fBstress\fP program is a Linux command-line utility not related to or
provided by EnergyMon.
The \fBnproc\fP program is part of \fBGNU coreutils\fP.
.SH "BUGS"
.LP
Report bugs upstream at <https://github.com/energymon/energymon>
.SH "SEE ALSO"
.BR energymon\-idle\-power (1)

RELEASES.md
# Release Notes

## [Unreleased]
### Changed
 * More pedantic formatting in man pages

## [v0.2.4] - 2018-06-01
### Changed
 * WattsUp implementations:
   * Wait up to 5 seconds for a good data packet (or fail) before completing initialization
   * Purge data buffers during initialization
   * Poll device at shorter intervals to increase energy value accuracy
   * Better handling of incomplete data packets
 * osp-polling: Allow polling thread to be interrupted during sleep when finish is requested
 * energymon-power-poller: Account for true elapsed time to produce more accurate power values

### Fixed
 * wattsup-libftdi, wattsup-libusb: Fix empty reads by always setting baud rate and serial line properties
 * wattsup-dev: Use pselect to avoid timer expired errors
 * odroid, odroid-ioctl, osp-polling, zcu102: Minor correction to polling delay interval
 * msr: Possible (but unlikely) precision computation bug
 * osp, osp-polling: Fix status check during init that didn't always work


## [v0.2.3] - 2017-11-19
### Added
 * Implementation for Xilinx ZCU102 (Jason Miller)

### Fixed
 * Fixed [#40]: Time conversion bug in common function 'energymon_gettime_elapsed_us' (Jason Miller)


## [v0.2.2] - 2017-10-26
### Added
 * New capabilities to some utility applications:
   * energymon-cmd-profile: Accept commands without quotations; add help option
   * energymon-file-provider: Add help, count, interval, and no-rewind options
   * energymon-idle-power: Add help option
 * Man pages for utility applications
 * VERSION and SOVERSION properties on shared object libraries
 * Multiarch support (use GNU standard installation directories)

### Changed
 * Increased minimum CMake version from 2.8 to 2.8.5 to support GNUInstallDirs
 * Refactor this RELEASES.md file

### Removed
 * Removed private symbol exports in shared object libraries (prevents symbol name clashes)

### Fixed
 * Various bug fixes to utility applications


## [v0.2.1] - 2017-10-01
### Added
 * Support for msr-safe kernel module in msr implementation
 * New cray-pm implementations to support capabilities of Cray XC30 and XC40 systems
 * New wattsup-libusb and wattsup-libftdi implementations
 * AppVeyor test configuration for Windows continuous integration

### Changed
 * Faster refresh in osp-polling, odroid and odroid-ioctl implementations
 * Better multi-platform support for some implementations, especially with portable time and sleep abstractions
 * Downgraded required CMake version from 2.8.9 to 2.8
 * Documented FSE 2016 and Technical Report publication information
 * Various enhancements, optimizations, and documentation improvements

### Deprecated
 * Deprecated custom DEFAULT CMake option in favor of ENERGYMON_BUILD_DEFAULT

### Removed
 * Removed Android CMake toolchain file: cmake-toolchain/android.toolchain.cmake

### Fixed
 * Fixed [#37]: In osp implementation, force overflow in ODROID Smart Power at 1000 Wh to avoid device counter saturation at 8192 Wh
 * Fixed [#27]: More modular build process
 * Numerous other bugs

## v0.2.0 - 2016-10-12
### Added
 * Initial public release

[Unreleased]: https://github.com/energymon/energymon/compare/v0.2.4...HEAD
[v0.2.4]: https://github.com/energymon/energymon/compare/v0.2.3...v0.2.4
[v0.2.3]: https://github.com/energymon/energymon/compare/v0.2.2...v0.2.3
[v0.2.2]: https://github.com/energymon/energymon/compare/v0.2.1...v0.2.2
[v0.2.1]: https://github.com/energymon/energymon/compare/v0.2.0...v0.2.1
[#40]: https://github.com/energymon/energymon/issues/40
[#37]: https://github.com/energymon/energymon/issues/37
[#27]: https://github.com/energymon/energymon/issues/27


utils/energymon-cmd-profile.c
/**
 * Capture the time/energy behavior of a system command.
 *
 * @author Connor Imes
 * @date 2016-03-02
 */
#define _GNU_SOURCE
#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include "energymon-default.h"
#include "energymon-time-util.h"

#define CMD_MAX_LEN 8192

static const char short_options[] = "+h";
static const struct option long_options[] = {
  {"help",      no_argument,       NULL, 'h'},
  {0, 0, 0, 0}
};

static void print_usage(int exit_code) {
  fprintf(exit_code ? stderr : stdout,
          "Usage: energymon-cmd-profile [OPTION]... COMMAND [ARG...]\n"
          "Options:\n"
          "  -h, --help               Print this message and exit\n");
  exit(exit_code);
}

int main(int argc, char** argv) {
  char cmd[CMD_MAX_LEN] = { 0 };
  energymon em;
  uint64_t time_start_ns;
  uint64_t time_end_ns;
  uint64_t time_total_ns;
  uint64_t energy_start_uj;
  uint64_t energy_end_uj;
  uint64_t energy_total_uj;
  double watts;
  int cmd_ret;
  int i;
  int written = 0;
  int c;
  while ((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
    switch (c) {
      case 'h':
        print_usage(0);
        break;
      case '?':
      default:
        print_usage(1);
        break;
    }
  }
  if (optind >= argc) {
    print_usage(1);
  }

  for (i = optind; i < argc; i++) {
    written += snprintf(cmd + written, CMD_MAX_LEN - written, "%s%s",
                        argv[i], (i == argc - 1 ? "" : " "));
    if (written >= CMD_MAX_LEN) {
      fprintf(stderr, "Command too long");
      exit(1);
    }
  }

  // initialize
  if (energymon_get_default(&em)) {
    perror("energymon_get_default");
    exit(1);
  }
  if (em.finit(&em)) {
    perror("energymon:finit");
    exit(1);
  }

  // get start time/energy
  errno = 0;
  energy_start_uj = em.fread(&em);
  if (!energy_start_uj && errno) {
    perror("energymon:fread");
    em.ffinish(&em);
    exit(1);
  }
  time_start_ns = energymon_gettime_ns();
  
  // execute
  printf("Executing: %s\n", cmd);
  cmd_ret = system(cmd);
  
  // get end time/energy
  errno = 0;
  energy_end_uj = em.fread(&em);
  if (!energy_end_uj && errno) {
    perror("energymon:fread");
    em.ffinish(&em);
    exit(1);
  }
  time_end_ns = energymon_gettime_ns();

  if (cmd_ret) {
    fprintf(stderr, "Warning: command exited with return code %d:\n%s\n",
            cmd_ret, cmd);
  }

  time_total_ns = time_end_ns - time_start_ns;
  energy_total_uj = energy_end_uj - energy_start_uj;
  watts = energy_total_uj * 1000.0 / time_total_ns;
  printf("Time (ns): %"PRIu64"\n", time_total_ns);
  printf("Energy (uJ): %"PRIu64"\n", energy_total_uj);
  printf("Power (W): %f\n", watts);

  // cleanup
  if (em.ffinish(&em)) {
    perror("energymon:ffinish");
  }

  // return the exit code of the command
  return cmd_ret;
}
